add josh's production checklist, slightly reorg self-hosting (#2961)
temporalio/documentation: docs/production-deployment/self-hosted-guide/checklist.mdx

  - versioning
---

> This page is intended to describe the most common challenges for customers who choose to self-host Temporal and share recommendations to mitigate those challenges. 

Reviewer: ```suggestion
This page describes common challenges customers face who self-host Temporal and it shares recommendations to mitigate those issues. 
```

> Temporal at its core is about durability and reliability.

Reviewer: I think this opening can be stronger and simpler. Something along the lines of: "At its core, Temporal delivers trusted deployment, built on reliability and durability. The following guide outlines how can verify your service reliability."

This durability is compromised if it is not architected and deployed as a reliable, trusted platform.
This guide provides a path to demonstrate that Temporal consumers can be confident in a Temporal service and provides a list of key tests you as a user should perform against the service, such as:

This guide provides a path to demonstrate that Temporal consumers can be confident in a Temporal service and provides a list of key tests you as a user should perform against the service.

## Self-Hosting Challenge Areas

> Significant engineering must be done and ongoing effort is required to resolve these challenges:

Reviewer: This feels like you started mid paragraph. Maybe include some opening words about what makes Self-hosting unique? Also phrasing like "must be done" and "It is critical to demonstrate confidence" can be reworked with some automated tooling.

- Scalability with spiky or growing workloads
- Global hosting
- Uptime, availability and reliability
- Management and control plane
- Latency, which must be kept low and consistent
- [Security](/self-hosted-guide/security)
- Maintenance and upgrades
- Expert support to users of the service
- Cost management

It is critical to demonstrate confidence in these areas as part of building a mission critical Temporal Service.
Without demonstrated architectural durability, the value of Temporal's [Durable Execution](https://temporal.io/how-it-works) model is compromised.

## Scalability with Variable or Growing Workloads {#scaling-and-metrics}

> Workloads can be highly variable, and sustained workload spikes of 5-10x average is common. 

Reviewer: Why are Workloads so variable. This again feels rushed, without introducing the fact that people use Workloads under various circumstances -- from Black Friday to a slow warm Sunday (not that I'm recommending those words)

Author: Good point -- removed some of the excessively detailed-yet-vague numbers here

> We recommend scaling your clusters to well above the average throughput.

Reviewer: I'd love to know if and when we can use "We". I know Marketing had some thoughts about this. Generally I ended up changing "We" to "Temporal" if I wasn't sure.

Reviewer: By the way, the advice is excellent.

Author: *Some* of our legacy docs have "we," and because these are recommendations, it's hard to remove it outright. Anyone we should @ here for a head check?

Author: Actually, removed after second round of feedback.

See [Scaling Temporal: The Basics](https://temporal.io/blog/scaling-temporal-the-basics) for an introduction to the topic.

> Temporal Server throughput is often limited by the number of [shards](/clusters#history-shard) configured for the Temporal service.

Reviewer: Service should be capped, and although shards is linked to a definition, a few words would be nice to explain it so people don't have to jump away.

Again, what I'm seeing is first rate advice and first draft wording. This is great stuff.

Author: Add one more definition sentence, ty!

  Use `workflow_success`, `workflow_failed`, `workflow_timeout`, `workflow_terminate` and `workflow_cancel` counters for each type of Workflow Execution completion.
  These include the `namespace` tag.

## Availability

A high level of availability and reliability (99.99%) is a requirement for mission critical deployments.
We recommend testing for this availability level while load testing.
We also recommend validating this level of reliability while doing server upgrades, to ensure no loss of service availability.  

Temporal Clusters can be deployed in as many regions as needed to meet various requirements:
- Data Residency
- Latency
- Security / Isolation
This can multiply the effort to implement and maintain clusters.

[Temporal Cloud is available in 15 AWS regions](/cloud/service-availability).

## Management and Control Plane 

> When you experience success with Temporal, you will typically expand your deployment.

Reviewer: Throwing this in, although it's so "marketing": "Temporal success leads to larger Temporal deployments. When you experience how well Temporal solves your problems and lowers costs, the natural response is..." Thought it might be inspirational as a comment.

  Use `workflow_success`, `workflow_failed`, `workflow_timeout`, `workflow_terminate` and `workflow_cancel` counters for each type of Workflow Execution completion.
  These include the `namespace` tag.

## Availability

A high level of availability and reliability (99.99%) is a requirement for mission critical deployments.
We recommend testing for this availability level while load testing.
We also recommend validating this level of reliability while doing server upgrades, to ensure no loss of service availability.  

Temporal Clusters can be deployed in as many regions as needed to meet various requirements:
- Data Residency
- Latency
- Security / Isolation
This can multiply the effort to implement and maintain clusters.

[Temporal Cloud is available in 15 AWS regions](/cloud/service-availability).

## Management and Control Plane 

Temporal success leads to larger Temporal deployments.
Needs can increase, and can go from having one or two production use cases in a single region to many use cases in many regions.

> Running multiple clusters is complex work, as each needs its own setup, tuning, and configuration. 

Reviewer: ```suggestion
Running multiple Temporal Services is complex work, as each needs its own setup, tuning, and configuration. 
```

> This operational management pain leads to a need to monitor and manage all of your Temporal Services in a unified way.

Reviewer: Maybe flip the last half and the first half of the sentence? Wording suggestions: 

> "Needing to monitor and manage all your Temporal Services in a unified way leads to operational management pain. Consider adding a layer on top of Temporal to manage multiple Temporal Services: a control plane. A control plane manages and directs data flow, deciding where data packets should be sent. A Temporal Service data plan can streamline operations and improve efficiency. Since Temporal does not ship its own open source data plane, rolling your own can be complex and take effort to add." 

But then there's no suggestions about how to find or do this or even where to look for solutions.

To do this, you need a layer on top of Temporal to manage multiple clusters: a control plane.
Temporal does not ship an open source control plane. This can be complex to do and take significant effort.

> With Temporal Cloud, all namespaces in all regions can be managed from a single view.

Reviewer: Cap Namespaces

The Temporal Cloud control plane allows operators to observe, manage, and control.

> Responsibilities for individual namespaces can be delegated using RBAC built into [Temporal Cloud](https://temporal.io/cloud).

Reviewer: Again, I'd flip the sentence halves. RBAC...can delegate...

> Self-hosted Temporal does not support RBAC or audit logging out of the box.

Reviewer: I'd flip this sentence and the next one, so the solution first and then the caveat (TCloud provides. Unfortunately, self-hosted doesn't)

Author: Keeping this one the way it is after making the previous change, because that way we're still ending the section on a positive note with Cloud.

Temporal Cloud provides RBAC and SSO support, audit logging, data encryption, third party penetration test validation, and SOC 2-Type II and HIPAA compliance.

## Maintenance and Upgrades 

> Temporal Server releases as often as every two weeks.

Reviewer: Missing a sentence before this. "Temporal Server is something something proactive maintained"

> Temporal recommends [upgrading sequentially](/self-hosted-guide/upgrade-server), not skipping any minor versions.

Author: Yeah, good point, switching to "Temporal recommends" across the board.

> It is acceptable to skip patch versions.

Reviewer: Maybe combine into previous ", although you can skip patch versions"?

Temporal recommends keeping up-to-date and not falling behind on your server versions.
No support is guaranteed for Temporal Server, but very old servers will be hard for even the community to support, so we encourage you to keep up to date. 

> In addition to the core service, infrastructure to run Temporal, such as Kubernetes to host the service and data stores for persistence, has to be implemented and maintained.

Reviewer: core Service? I _think_? capped.

Also, this sentence can be de-mangled a little:

> "You must create and maintain the infrastructure to host and run your self-hosted Temporal installation (for example, Kubernetes) and data stores for persistence."

I know that we're not supposed to use parentheticals and I expect you can write this better. I just wanted to note that the sentence is a little hard to read and understand in its current form.

Author: Oh yeah, this sentence was pretty hard to read, thanks for identifying this. Went with:

"You must create and maintain the infrastructure to host and run your self-hosted Temporal installation, such as Kubernetes, as well as data stores for persistence."

> Temporal service availability can be affected by server upgrades.

Reviewer: service cap. Also flip the sentence "Server upgrades can negatively affect Temporal Service availability" or something like that

> We recommend load and availability testing during the upgrade process to understand the performance envelope.

Reviewer: "performance envelope" may be a little too jargony without softening with a tiny bit of explanation. OpenAI says: "the range of conditions within which a system, component, or device can operate effectively and safely"

Author: Just going to go with "performance implications," I think.

Temporal Cloud updates are managed by the Temporal Cloud team.

> Cloud upgrades are seamless.

Author: Semicolon time!

## Expert Support

Temporal recommends that customer platform teams who are building out a Temporal service gain deep experience across the lifecycle and breadth of a Temporal application.

Specific activities include:
- [Worker tuning](/develop/worker-performance)
- [Worker best practices](/workers)
- Code reviews
- Design guidance
- Training
- Code reviews
- Security reviews
- [Metrics](/references/sdk-metrics) and monitoring
- Technical onboarding

> Temporal support provides guidance on all of the above.

Reviewer: If we provide guidance for all of the above, then should this link for every bullet point?

Author: I think no -- because these aren't all necessarily referring to publicly available documentation, but categories of support we offer to customers.

> Temporal support will work with your teams to make this easier and support teams that support Temporal.

Reviewer: The word support is used 3 times in this sentence. Also, whenever mentioning Temporal support, I normally add a link to [file a ticket with Temporal Support] and work it in there somehow, so it's right at the point where people might be when they're trying to work through a solution

Author: Good point -- I'm just going to remove this redundant sentence, I think adding a link to the Support page to the previous sentence is sufficient.

## Cost Management

Running a mission critical, global Temporal Service can be expensive.
Temporal Server is a complex system to run and scale.

> We recommend performance testing as outlined above, and planning scaling around your performance requirements.

Reviewer: I don't think this needs to refer back (as outlined above) and I think the second half is a little hard to understand "planning scaling" two ings in a row. Maybe planning how you'd scale as your performance requirements evolve?

This implies that Temporal Servers will often be over-sized, but this is necessary to handle unpredictable spiky workloads.
Performance testing can help you right-size your environments.

> Running mission critical Temporal as a Service requires multiple Temporal Clusters for high availability and global coverage. 

Reviewer: Temporal Service instances instead of Clusters

Author: This might actually be one of the cases where we continue referring to Clusters (both because it's somewhat more specific to deployment scaling, and because having "Temporal Service" in the same sentence as "Temporal as a Service" is kind of gnarly.

> It is a good practice to have trained, experienced administrators familiar with Temporal server architecture to maintain Temporal servers, with global coverage to maintain a global, mission critical service.

Reviewer: Active: to train seasoned administrators familiar with. Should Server be capped here? Maybe? 
Maintain _your_ self-hosted Temporal servers and use the global (capped?) coverage features to maintain a ...

Love the emphasis on mission-critical service.

Author: replaced the first mention of server with Service, I think that helped:

"It is a good practice to have trained, experienced administrators familiar with Temporal Service architecture to maintain your Temporal servers and provide a mission critical service."

> Staffing, training and skill development can be significant costs to maintaining a Temporal Service.

Reviewer: This is so negative unless you contextualize it right away. I'd suggest you soften this by moving the Temporal cloud line (currently 157 but it might not stay that way) up to just before "It's a good practice"

Author: See my other comment about wanting to *end* on the positive note of the comparison -- I think bringing it back to Temporal Cloud at the end here leaves a(n intentionally) less negative impression!
